File: tests/julia/test_eval.jl
# test_eval.jl
# Tests for expression evaluation (User Story 1)

include(joinpath(@__DIR__, "load_wrapper.jl"))

@testset "GIAC Wrapper Eval Tests" begin
    @testset "Version Functions" begin
        # giac_version() returns CxxWrap.StdLib.StdStringAllocated, not a
        # Base.String; AbstractString accepts both.
        @test giac_version() isa AbstractString
        @test !isempty(giac_version())
        # Issue #2: wrapper_version() now reads from meson.project_version()
        # at build time, so it never falls out of sync with meson.build.
        # Accept any SemVer-shaped string instead of a hardcoded literal.
        @test occursin(r"^\d+\.\d+\.\d+", wrapper_version())
    end

    @testset "GIAC Availability" begin
        @test is_giac_available() == true
    end

    @testset "Basic Eval" begin
        ctx = GiacContext()
        @test giac_eval(ctx, "1+1") == "2"
        @test giac_eval(ctx, "2*3") == "6"
    end

    @testset "Factor Operation" begin
        ctx = GiacContext()
        result = giac_eval(ctx, "factor(x^2-1)")
        @test occursin("x-1", result) || occursin("x+1", result)
    end

    @testset "Error Handling" begin
        # GIAC's parser is permissive: `invalid(((` triggers a warning and
        # returns `undef`, it does NOT throw. `factor()` with no argument
        # does throw, so use it as the exception witness.
        ctx = GiacContext()
        @test_throws Exception giac_eval(ctx, "factor()")
    end
end

File: tests/cpp/test_eval.cpp
/**
 * @file test_eval.cpp
 * @brief Tests for expression evaluation (User Story 1)
 */

#include "giac_impl.h"
#include <iostream>
#include <cassert>
#include <string>

using namespace giac_julia;

// Simple test framework macros
#define TEST(name) void test_##name()
#define RUN_TEST(name) do { \
    std::cout << "Running " #name "... "; \
    try { test_##name(); std::cout << "PASSED" << std::endl; } \
    catch (const std::exception& e) { std::cout << "FAILED: " << e.what() << std::endl; return 1; } \
} while(0)

#define ASSERT_EQ(expected, actual) do { \
    if ((expected) != (actual)) { \
        throw std::runtime_error("Expected: " + std::string(expected) + ", Got: " + std::string(actual)); \
    } \
} while(0)

#define ASSERT_THROWS(expr, exception_type) do { \
    bool caught = false; \
    try { expr; } catch (const exception_type&) { caught = true; } \
    if (!caught) throw std::runtime_error("Expected exception not thrown"); \
} while(0)

// T016: Test basic eval - "1+1" returns "2"
TEST(basic_eval) {
    GiacContext ctx;
    std::string result = ctx.eval("1+1");
    ASSERT_EQ("2", result);
}

// T017: Test factor operation - "factor(x^2-1)"
TEST(factor_operation) {
    GiacContext ctx;
    std::string result = ctx.eval("factor(x^2-1)");
    // Result should be "(x-1)*(x+1)" or equivalent
    // GIAC may format differently, so we evaluate back
    ASSERT_EQ("(x-1)*(x+1)", result);
}

// T018: Test error handling - invalid expression
TEST(error_handling) {
    GiacContext ctx;
    ASSERT_THROWS(ctx.eval("invalid((("), std::runtime_error);
}

// Additional tests
TEST(version_functions) {
    std::string gv = get_giac_version();
    std::string wv = get_wrapper_version();
    ASSERT_EQ("0.1.0", wv);
    // GIAC version should be non-empty
    assert(!gv.empty());
}

TEST(giac_available) {
    assert(check_giac_available());
}

int main() {
    std::cout << "=== GIAC Wrapper Eval Tests ===" << std::endl;

    RUN_TEST(version_functions);
    RUN_TEST(giac_available);
    RUN_TEST(basic_eval);
    RUN_TEST(factor_operation);
    RUN_TEST(error_handling);

    std::cout << "=== All tests passed ===" << std::endl;
    return 0;
}
